[DNM][WIP][do not review yet]Service and Device initialization: automatic priorities

tbursztyka opened the pull request:
DO NOT REVIEW YET (I just need to see first what CI will find out, my potato PC cannot run the whole test suites)

This is a fix for #76937  after discussing on the RFC #73836

I applied most comments made during these discussion. 
- Keeping current macros unchanged and able to live with the new ones
- Adding new macros: both for device and service (not init, SYS_INIT() will be deprecated as well anyway). Taking the opportunity to reduce the number of macros. Basically merging DT/non-DT based macros, so removing manually introduced device name. This is a legacy thing that needs to disappear. Explicit device name is coming from the time DTS was not used, and during the process of switching to DTS we had to keep it. Now it's not true anymore. The new macros still can instantiate a non-DT device, but it will take the unique token as a name and that's it.
- Declaring the explicit dependency resolution file to cmake via cmake functions and not by looking up on pre-determined places (only on the application side, I could not use such cmake function, so there it has to be a specific file, unless I misunderstood something with the build process)
- The new service API is extremely minimal as it is meant to just replace SYS_INIT where relevant (like, say: net_init, and not all the places where SYS_INIT should not have been used). BUT: it will grow, there are plenty of rooms for improvements. We could start/stop/pause a service if needed, etc etc...

Added to this I (tried to) switch most device macro calls to new ones. Which make this PR a huge one in files touched and modified lines. There are few corner cases which prevent changes: PTP and ethernet's DSA. These needs to switch to DTS.

Last but not least: brace for impact  /o\

tbursztyka committed:
include: Add init utility macro specific header

It provides useful macros to manipulate generated ZINIT_ macros from
gen_init_priorities.py script.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
cmake: Add a gen_init_priorties module

Run after Kconfig and DTS, so it can build up on top of the edt object.
Thus filling in any gap left by Kconfig and DTS.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
init: Adapt init entry section generation

The sub-prio is entirely removed, as the new macros will bring uniquely
computed priorities.

This is seamless: none of this is of public API, so no deprecation and
macro replacement is necessary at this stage.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
scripts/edtlib: Add the same accessor as chosen but for aliases

This will prove to be useful to retrieve the original DT Node from an
alias.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
linker: Add an extra number to check for sorting levels sections

Now that priorities are automatically computed, and mostly coming from
DTS, it can easily go beyond 99.

Actually, this arbitrary limit of 99 in level sections has no real
meaning anymore (if it ever had any).

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
init: Add an extra '_' between level names and priority in sections

This is an optional commit, but it makes things slightly easier to read
in zephyr.map.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
devicetree: Define DT_NODE_HAS_STATUS if not ZTEST_UNITTEST

linker-defs.h does a weird trick around DT_NODE_HAS_STATUS. The issue is
that intrications of includes might anyway end up including
linker-defs.h and devicetree.h (as it requires other macros in that last
header).

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
scripts: Add Zephyr initalization object script

This script fills in the gap between Kconfig and DTS to generate a
header allowing to remove all manual priority parameter that are present
on SYS_INIT() and DEVICE_<DT_>DEFINE based macros, by automatically
generating the right values from DTS and from any given service or
device yaml file.
New macros will take advantage of this and will be replacing SYS_INIT
and DEVICE_<DT>_DEFINE ones.

This will permit to fully relate the software services and the devices.
For instance, it is now possible to make uart_console_init
depending on the actual uart device to be initialized first.
And, thanks to DTS, this can be generalized via using "zephyr,console"
alias. Such as:

service:
    uart_console_init:
        init:
            level: POST_KERNEL
            dependencies:
                - zephyr,console

Note that level property can be tweaked as well, and if so: will
supersede the one given to the macro which is interpreted as a default.
Such level becomes an EXPLICIT one.
An addition to that, any other node depending on a node with an explicit
level, will see its level implicitely set: its level become an
IMPLICIT one.

Behavior will be so that:
- explicit level will replace the default unconditionally
- implicit level will replace the default if it is a higher one only.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
service: Introduce software service instanciation macros

SERVICE_INSTANCE_INIT_ONCE is a direct replacement for SYS_INIT_NAMED,
removing the priority parameter as it is now provided by
gen_init_priorities.py script.

No alias will be created for SYS_INIT as we seek to limit the usage of
SERVICE_INSTANCE_INIT_ONCE, on the contrary to SYS_INIT which got widely
abused in the past.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
device: Add new macros to instanciate device objects

In order to get rid of the manually inserted initialization priority,
device object generation need new macros.

Until now, 2 set of macros have co-existed because of legacy support.
Indeed, devices used to be created fully manually before devicetree
introduction. Once devicetree appeared, DT based macros were added.
Now is the time to get rid of such difference, and have a unique macro
for creating device objects.

So instead of DEVICE_DEFINE() and DEVICE_DT_DEFINE() there will be
DEVICE_INSTANCE(). Instead of DEVICE_DT_INST_DEFINE() there will be
DEVICE_INSTANCE_FROM_DT_INST(). Instead of DEVICE_GET() and
DEVICE_DT_GET(), there will be DEVICE_INSTANCE_GET().
Same for DEVICE_INIT_DT_GET() and DEVICE_INIT_GET(), there will be
DEVICE_INIT_ENTRY_GET()

In the future step of deprecating old macros, these are then identified:
- DEVICE_DEFINE
- DEVICE_DT_DEFINE
- DEVICE_GET
- DEVICE_DT_GET
- DEVICE_INIT_DT_GET
- DEVICE_INIT_GET
- DEVICE_NAME_GET() could be replaced by DEVICE_INSTANCE_NAME())
- DEVICE_DT_NAME_GET() could be replaced by DEVICE_INSTANCE_NAME())
- DEVICE_DECLARE() could be replaced as well

And probably some more, thus reducing the amount of macros which would
help to clean-up the header and have a more concise set.

The difference is now that priority parameter is removed, as well as the
name parameter (originaly from DEVICE_DEFINE legacy macro). In this case
it is just simpler to reuse the "unique token" dev_id.

Because of the dependency being fully computed _prior_ to building
zephyr, these new macros do not support CONFIG_DEVICE_DEPS anymore. This
feature was never really used anywhere and is a legacy thing. So new
macros do not accept variable arguments.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
CMakeLists.txt: call zephyr_generate_init_priorities() where relevant

All service or device init files added to any valid CMakeLists.txt will
now be used to generate include/zephyr/generated/zinit.h header file,
with computed priorities.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
sensor: Provide new device instantiation macros

In order to get rid of the manually inserted initialization priority,
device object generation need new macros.

Again removing the prio parameter.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
tests: Switching DEVICE_<DT/DT_INST>_DEFINE macros to new ones

Switching to DEVICE_INSTANCE and DEVICE_INSTANCE_FROM_DT_INST, thus
merging former DEVICE_DEFINE and DEVICE_DT_DEFINE into one, so removing
the name parameter from the earier and finally removing the prio
parameter on all.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
subsys: Switching DEVICE_<DT/DT_INST>_DEFINE macros to new ones

Switching to DEVICE_INSTANCE and DEVICE_INSTANCE_FROM_DT_INST, thus
merging former DEVICE_DEFINE and DEVICE_DT_DEFINE into one, so removing
the name parameter from the earier and finally removing the prio
parameter on all.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
boards: Switching DEVICE_<DT/DT_INST>_DEFINE macros to new ones

Switching to DEVICE_INSTANCE and DEVICE_INSTANCE_FROM_DT_INST, thus
merging former DEVICE_DEFINE and DEVICE_DT_DEFINE into one, so removing
the name parameter from the earier and finally removing the prio
parameter on all.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
sensors: Switching SENSOR_DEVICE_<DT/DT_INST>_DEFINE macros to new ones

Switching to SENSOR_DEVICE_INSTANCE and
SENSOR_DEVICE_INSTANCE_FROM_DT_INST, thus removing the prio parameter
on all.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
drivers: Switching DEVICE_<DT/DT_INST>_DEFINE macros to new ones

Switching to DEVICE_INSTANCE and DEVICE_INSTANCE_FROM_DT_INST, thus
merging former DEVICE_DEFINE and DEVICE_DT_DEFINE into one, so removing
the name parameter from the earier and finally removing the prio
parameter on all.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
net: Provide new device instantiation macros

Again removing the prio parameter.

Managing the ident as a DT node-id or dev_id helps to reduce the number
from 12 to 7 macros.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
drivers: Switching NET_DEVICE_<INIT/DT_DEFINE>... macros to new ones

Switching to NET_DEVICE_INSTANCE, NET_DEVICE_INSTANCE_FROM_DT_INST,
NET_DEVICE_INSTANCE_MULTI, NET_DEVICE_INSTANCE_MULT_FROM_DT_INST,
NET_DEVICE_OFFLOAD_INSTANCE and
NET_DEVICE_OFFLOAD_INSTANCE_FROM_DT_INST.

Thus removing prio and name parameters on all.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
net/ethernet: Provide new device instantiation macro

Again removing the prio and name parameters.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
ethernet: Switching ETH_NET_DEVICE_<INIT/DT_INSTANCE>... to new ones

Switching to ETH_NET_DEVICE_INSTANCE,
ETH_NET_DEVICE_INSTANCE_FROM_DT_INST and ETH_NET_DEVICE_INSTANCE_MULTI.

Thus removing prio and name parameters on all.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
tests: Switching NET_DEVICE_<INIT/DT_DEFINE>... macros to new ones

Switching to NET_DEVICE_INSTANCE, NET_DEVICE_INSTANCE_FROM_DT_INST,
NET_DEVICE_INSTANCE_MULTI, NET_DEVICE_INSTANCE_MULT_FROM_DT_INST,
NET_DEVICE_OFFLOAD_INSTANCE and
NET_DEVICE_OFFLOAD_INSTANCE_FROM_DT_INST.

Thus removing prio and name parameters on all.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
can: Provide new device instantiation macro

Again removing the prio parameter.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
drivers/can: Switching CAN_DEVICE_DT_DEFINE macros to new ones

Switching to CAN_DEVICE_INSTANCE and CAN_DEVICE_INSTANCE_FROM_DT_INST
relevantly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
smbus: Provide new device instantiation macro

Again removing the prio parameter.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
drivers/smbus: Switching SMBUS_DEVICE_DT_DEFINE macros to new ones

Switching to SMBUS_DEVICE_INSTANCE and
SMBUS_DEVICE_INSTANCE_FROM_DT_INST relevantly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
include/drivers: Switching specific headers to new device macros

Lone files that define their own wrapper macro on top of formely
DEVICE_DT_DEFINE macros.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
i2c: Provide new device instantiation macro

Again removing the prio parameter.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
drivers/i2c: Switching I2C_DEVICE_DT_DEFINE macros to new ones

Switching to I2C_DEVICE_INSTANCE and I2C_DEVICE_INSTANCE_FROM_DT_INST
relevantly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
spi: Provide new device instantiation macro

Again removing the prio parameter.

However, original macro does not seem to be used anywhere...

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
firmware: Provide new device instantiation macro

Again removing the prio parameter.

And switching the unique macro call in drivers.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
tests and samples: Add relevant service and device init modifiers

A few test and samples require explicit dependency resolution as the
devices they introduce do not have any DTS counter-part.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

cfriedt commented:
@tbursztyka - it might be worth it to put the PR into draft state until you feel it's ready for review. Thanks for taking the initiative to tackle this issue 👍

tbursztyka commented:
> @tbursztyka - it might be worth it to put the PR into draft state until you feel it's ready for review. 

good point, and I'll remve the DNM label so CI will go through more tests.

tbursztyka committed:
drivers: Add relevant changes for dependency resolution

A few device drivers need to get the dependency properly set up.
This is only true for non-DT based devices.

One of the changes being the device "unike token" which was made of
capital letter for some reason on crypto side etc...

Basically, legcay things that should disappear and be fully managed at
DTS level.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
tests: Fix DTS overlay for build_all/input

It's clearly a bug: an input device is not meant to listen to itself.
This creates a dependency loop, and it is actually surprising this did
not get detected by DTS (which does call add_edge on graph on the same
node).

Binding description of the node says 'input' property can be left empty
so removing the line entirely.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
modules: Switching DEVICE_<DT/DT_INST>_DEFINE macros to new ones

Switching to DEVICE_INSTANCE and DEVICE_INSTANCE_FROM_DT_INST, thus
merging former DEVICE_DEFINE and DEVICE_DT_DEFINE into one, so removing
the name parameter from the earier and finally removing the prio
parameter on all.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
soc: Switching DEVICE_<DT/DT_INST>_DEFINE macros to new ones

Switching to DEVICE_INSTANCE and DEVICE_INSTANCE_FROM_DT_INST, thus
merging former DEVICE_DEFINE and DEVICE_DT_DEFINE into one, so removing
the name parameter from the earier and finally removing the prio
parameter on all.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>

tbursztyka committed:
tests: Change check_init_priorities fake device's priorities relevantly

Now that init ordering got back to a unique number (automatically
computed or provided manually with legacy macros) the priorities need to
be sorted relevantly so the erroneous ordering still matches the
expectations.

When legacy macros will be removed after a deprecation period, such test
and most probably check_init_priorities.py tool as well won't make sense
anymore.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@proton.me>
